src/sync_key_gen.rs

```diff
@@ -0,0 +1,238 @@
+//! A _synchronous_ algorithm for dealerless distributed key generation.
+//!
+//! This protocol is meant to run in a _completely synchronous_ setting where each node handles all
+//! messages in the same order. This can be achieved by making its messages transactions on top of
+//! `HoneyBadger`, or by running it "on-chain", i.e. committing its messages to a blockchain.
+//!
+//! Its messages are encrypted where necessary, so they can be publicly broadcast.
+//!
+//! When the protocol completes, every node receives a secret key share suitable for threshold
+//! signatures and encryption. The secret master key is not known by anyone. The protocol succeeds
+//! if up to `threshold` nodes are faulty.
+//!
+//! # How it works
+//!
+//! The algorithm is based on ideas from
+//! [Distributed Key Generation in the Wild](https://eprint.iacr.org/2012/377.pdf) and
+//! [A robust threshold elliptic curve digital signature providing a new verifiable secret sharing scheme](https://www.researchgate.net/profile/Ihab_Ali/publication/4205262_A_robust_threshold_elliptic_curve_digital_signature_providing_a_new_verifiable_secret_sharing_scheme/links/02e7e538f15726323a000000/A-robust-threshold-elliptic-curve-digital-signature-providing-a-new-verifiable-secret-sharing-scheme.pdf?origin=publication_detail).
+//!
+//! If there were a trusted dealer, they would generate a `BivarPoly` of degree `t` and publish
+//! the `BivarCommitment`, with which the polynomial's values can be publicly verified. They'd
+//! then send _row_ `m > 0` to node number `m`. Node `m`, in turn, sends _value_ `s` to node number
+//! `s`. Then if `2 * t + 1` nodes confirm that they received a valid row, and there are at most
+//! `t` faulty nodes, then at least `t + 1` honest nodes sent on an entry of every other node's
+//! column to that node. So we know that every node can now reconstruct its column and the value at
+//! `0` of its column. These values all lie on a univariate polynomial of degree `t`, so they can
+//! be used as secret keys.
+//!
+//! To avoid trusting a single dealer, we make sure that at least `t + 1` nodes use the above
```

Comment:
In our **dealerless** environment, at least `t+1` nodes each generate a polynomial using the method above. The sum of the secret keys from each node is used as the master secret key. No single node knows the sum of the secret master key.